packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -35,6 +36,25 @@ const int _kPubExitCodeUnavailable = 69;
 
 typedef MessageFilter = String? Function(String message);
 
+/// targetPath is the directory in which the content of the extraPath will be moved in
+void joinCaches(FileSystem fileSystem, String targetPath, String extraPath) {
+  final Directory targetDirectory = fileSystem.directory(targetPath);
+  final Directory extraDirectory = fileSystem.directory(extraPath);
+
+  for (final FileSystemEntity entity in extraDirectory.listSync()) {
+    final File file = fileSystem.file(entity.path);
```

Comment:
A `FileSystemEntity` doesn't have to be a file, I think instead you need to check if `entity is File`